### ISSUE TITLE
SHAP interactions

### ISSUE THREAD

**Issue by mayer79** [microsoft/LightGBM#6058]
## Summary

Scott Lundberg has added SHAP interactions into his TreeSHAP implementation in XGBoost, but not in LightGBM. This option is very helpful in opening the "black box". I'd love to see SHAP interaction values also in LightGBM. 

There are a couple of TreeSHAP experts around @hbaniecki that might be willing to implement it directly into LightGBM.

## Description

SHAP interactions decompose raw predictions into the sum of contributions from all feature pairs. There is an implementation in C++ in XGBoost (by Scott, I think), and there is the R package {treeshap} https://github.com/ModelOriented/treeshap with another implementation in C++ that also works for LightGBM. 

## References

Lundberg, S.M., Erion, G., Chen, H. et al. From local explanations to global understanding with explainable AI for trees. Nat Mach Intell 2, 56–67 (2020). https://doi.org/10.1038/s42256-019-0138-9

**Comment by jameslamb** [comment]
Linking some related conversations:

* #3127
* https://github.com/microsoft/LightGBM/issues/1350#issuecomment-395903942

**Comment by mayer79** [comment]
Here the implementation in {treeshap}:

https://github.com/ModelOriented/treeshap/blob/master/src/treeshap.cpp

**Comment by kiparkmoon** [comment]
@mayer79 
Thanks for your kindness.
Your statement is like shap interaction value didn't work in light gbm(python). However, there is a way to use it(C++ and  R).
Does it mean that if i want to get the shap interaction value using lightgbm, then trying another process except python?(actually my dataset have several caterogical variables).

**Comment by mayer79** [comment]
@kiparkmoon not exactly. Lightgbm and XGBoost contain TreeSHAP as part of the library. XGBoost additionally provides SHAP interactions. 

It would be fantastic to have SHAP interactions also as part of the lightgbm library.

**Comment by kiparkmoon** [comment]
@mayer79  I've tried SHAP interaction values for both XGBoost and LightGBM now, and it seems to work SHAP interaction values as long as there are no categorical variables.  I pulled out the SHAP interaction value for those.

**Comment by mayer79** [comment]
In Python, this is done outside the Lightgbm library, namely directly in {shap}. The aim of this issue is to bring it into lightgbm, so that it would also be available outside Python.

**Comment by hjjhh6** [comment]
My comment is to support the implementation of SHAP interaction value calculation in LightGBM in Python. Considering that the SHAP library has yet to implement SHAP interaction value calculation when LightGBM uses "categorical_feature" as of 2026 (LightGBM version: 4.6.0, SHAP version: 0.48.0). Consequently, I had to migrate a significant portion of my scientific work from LightGBM to XGBoost. Therefore, implementing SHAP interaction value calculations for categorical features within the LightGBM library is critically important in my experience. This represents a key advantage XGBoost holds for scientific research.

**Comment by hbaniecki** [comment]
Hi, feel free to check out our `shapiq` Python package for explaining interactions https://github.com/mmschlk/shapiq albeit not much can be done in terms of categorical features.

**Comment by mayer79** [comment]
@hbaniecki it is all about categoricals... even if shapiq is very neat!
